### ISSUE TITLE
Issues in User32.GetWindowText() (out-of-bounds write; text length change detection not working)

### ISSUE THREAD

**Issue by kpreisser** [dotnet/winforms#2424]
* .NET Core Version: `master` (e0a1361874fce56f7a4fac2ec527e155e7c76128)

**Problem description:**
While working on the task dialog PR, I had a look at the implementation of `User32.GetWindowText()` and noticed it seems to have some issues:

https://github.com/dotnet/winforms/blob/e0a1361874fce56f7a4fac2ec527e155e7c76128/src/Common/src/Interop/User32/Interop.GetWindowText.cs#L26-L53

* When calling `GetWindowTextW`, the `nMaxCount` parameter is specified as `textLengthWithNullTerminator + 1`, but the call to `ArrayPool<char>.Shared.Rent` specifies `textLengthWithNullTerminator`.
`ArrayPool.Rent` returns a buffer that has at least the specified size, so it might return a buffer that has exact the specified size, in which case it will be 1 char too small to hold the maximum number of chars. This can lead to too many bytes being written (out-of-bounds write) if the text changes after calling `GetWindowTextLengthW`. <br>
For example, consider the following scenario:
  - The control text is initially `"AAAABBBBCCCCDDD"` (15 chars).
  - `GetWindowTextLengthW(hWnd)` returns `15`, so `textLengthWithNullTerminator` is `16`.
  - `ArrayPool<char>.Shared.Rent(textLengthWithNullTerminator)` returns an array with length `16`. 
  - Meanwhile, the control text changes to `"AAAABBBBCCCCDDDx"` (16 chars).
  - `GetWindowTextW()` is called with `nMaxCount`=`17`. This parameter specifies the maximum buffer size/characers to write, including the terminating null character.
  - As the text has changed to 16 characters, `GetWindowTextW()` writes 17 characters (`"AAAABBBBCCCCDDDx\0"`) to the buffer and returns the value `16`, thus having written one char (2 bytes) too many. <br>
  The issue can be fixed by changing `ArrayPool<char>.Shared.Rent(textLengthWithNullTerminator)` to `ArrayPool<char>.Shared.Rent(textLengthWithNullTerminator + 1)`.<br><br>
* The implementation wants to handle the case when the text length increases between calling `GetWindowTextLengthW` and calling `GetWindowTextW`. However, the condition actually doesn't work, and a truncated text is returned in that case if the text length has increased by 2 or more chars.
  <br>Consider the following scenario:
  - The control text is initially `"A"` (1 chars).
  - `GetWindowTextLengthW(hWnd)` returns `1`, so `textLengthWithNullTerminator` is `2`.
  - Meanwhile, the control text changes to `"ABC"` (3 chars).
  - `GetWindowTextW()` is called with `nMaxCount`=`3` and therefore writes `"AB\0"` to the buffer and returns `2`.
  - The condition `actualTextLength > textLengthWithNullTerminator` evaluates to `false` because `actualTextLength==2` and `textLengthWithNullTerminator==2`.
  - The string `"AB"` is returned, whereas the correct value would be either `"A"` or `"ABC"` (depending on the timing).<br>
  The issue can be fixed by changing 
  `if (actualTextLength > textLengthWithNullTerminator)` to 
  `if (actualTextLength > textLengthWithNullTerminator - 1)`.<br><br>
* If the code determines that the text length has increased between the call to `GetWindowTextLengthW` and `GetWindowTextW`, the loop starts again with determining the new length by calling `GetWindowTextLengthW` (but doesn't keep the last length as minimum value).
  However, if the text changed again to a shorter text between the last call to `GetWindowTextW` and the call to `GetWindowTextLengthW`, `textLengthWithNullTerminator` will have the short value again, and if now the text changes again to a longer text and so forth in exact the same rythm, there can be an infinite loop (although that will be very unlikely in practice).<br>
  I think that the length variable (`textLengthWithNullTerminator`) should be set to the maximum of the previous `actualTextLength+1` and the new length returned by `GetWindowTextLengthW` to avoid this (so that the length will always increase in a new interation).

Another minor optimization would be to specify the actual buffer size (`windowTitleBuffer.Length`) rather than `textLengthWithNullTerminator + 1` in the call to `GetWindowTextW`, because the returned buffer might actually be longer than `textLengthWithNullTerminator + 1`, so that we can avoid unnecessary loops if the text length increases between the calls but the new text would still fit into the buffer.
In that case, the condition `if (actualTextLength > textLengthWithNullTerminator - 1)` would also need to be changed to `if (actualTextLength > windowTitleBuffer.Length - 2)`.

Note: It seems the buffer out-of-bounds issue (and the broken condition) were introduced with commit 162cba46ddbe80c2c1759cbcd3bbafbbf19f3492 (#923).

Thanks!

**Comment by RussKie** [comment]
Are you able to create a repro?

Looking at the code, one of callsites is wrapped with `MultithreadSafeCallScope`, which as I understand it prevent information alteration whilst we're in call. E.g. 
https://github.com/dotnet/winforms/blob/7aeb736b075b102efb4d3fc4df6add660177d032/System.Windows.Forms/src/System/Windows/Forms/Control.cs#L4036-L4039

However few other callsites aren't wrapped, and perhaps that is we need to apply a fix.

E.g.
* https://github.com/dotnet/winforms/blob/07ae71178eda9d1e4ebf73be44879d969cea7dc5/src/System.Windows.Forms.Design.Editors/src/System/ComponentModel/Design/MultilineStringEditor.cs#L362-L366
though, arguably this is a dialog and a chance of `Text` changing during the call is negligent
* https://github.com/dotnet/winforms/blob/69a16842ce6b77ff92f3ff547489abd4b551065a/src/System.Windows.Forms/src/System/Windows/Forms/FileDialog.cs#L129
same as above
* https://github.com/dotnet/winforms/blob/9bdd2e4e73ed5e9dd37e6e09a7d38143e64ec025/src/System.Windows.Forms.Design/src/System/ComponentModel/Design/DesignerActionUI.cs#L1003
this is a VS Designer-specific implementation
* https://github.com/dotnet/winforms/blob/716397fe0a1a60c77b2e68fa8cab5b1b4b6d6eba/src/System.Windows.Forms/src/System/Windows/Forms/WinFormsUtils.cs#L175-L177
this is only DEBUG specific

**Comment by weltkante** [comment]
> Are you able to create a repro?

I don't think requesting a repro is the right approach to this kind of bug, this is an out-of-bounds write, you shouldn't require a reproducable crash to fix those as it can be a potential security issue.

The problem is the code is lying about the buffer size you are passing, the buffer is one char smaller than its claiming to be, so GetWindowTextLengthW will do an out-of-bounds write without hesitating because the code is telling it to do so.

I don't think this is testable either, getting buffer sizes wrong is a risk you take when working with spans manually, you can always get an off-by-one error in there by accident.

**Comment by weltkante** [comment]
[edit] Never mind, I misunderstood some of the bugreport. The off-by-one error in the win32 API (`GetWindowTextW` asking for a larger size than it has buffer for) should be fixed in any case regardless of what is done with the rest of the reported issue.

**Comment by kpreisser** [comment]
> Are you able to create a repro?

As @weltkante noted earlier, it's probably not possible to create a repro for the first issue (out-of-bounds write) that causes a crash or throws an exception, as it's only one char that is written outside the bounds of the buffer which causes no visible effects in most cases. So I think the issue and fix will need to be verified manually.

Additionally, all of mentioned issues will only be obversable if the text length of the window/control changes while the code is executed, which I think is only possible when the window handle belongs to a different thread (or even process). So, a repro would need to synchronize the specific timing between the threads, which I don't think is possible without altering the code specifically for this.

However, it's possible to reproduce these issues manually, by running Visual Studio and debug the application. For example, to reproduce the second issue (truncated text), I used the following steps:

* Create a new project with `dotnet new console`.
* Add a `<ProjectReference>` to `System.Windows.Forms.csproj`.
* In project `System.Windows.Forms`, add a class like this allow `Interop.User32.GetWindowText` being accessed by the other project:
```c#
public static class Test_Issue242
{
    public static string GetWindowText(IntPtr hWnd) => Interop.User32.GetWindowText(hWnd);
}
```

* In method `Interop.User32.GetWindowText`, add the following code after the line with `ArrayPool<char>.Shared.Rent`, which makes it easier to determine how many characters were actually written by `GetWindowText`, and which adds a few instruction to make it easier to get the right timing while single-stepping in the debugger:
```c#
        Array.Fill(windowTitleBuffer, 'X');
        for (int i = 0; i < 5; i++)
            ;
```
* In the new project's `Program.cs`, use the following code (this program creates a form with a button whose text is initially `"A"` and later changed to `"ABC"`):
```c#
using System;
using System.Diagnostics;
using System.Threading;
using System.Windows.Forms;

namespace repro2424
{
    class Program
    {
        [STAThread]
        static void Main()
        {
            const string initialText = "A";
            const string finalText = "ABC";

            Application.EnableVisualStyles();
            Application.SetCompatibleTextRenderingDefault(false);

            var form = new Form();
            var button = new Button()
            {
                Text = initialText,
                FlatStyle = FlatStyle.System
            };
            form.Controls.Add(button);

            form.Load += (s, e) =>
            {
                var btnHandle = button.Handle;
                new Thread(() =>
                {
                    Thread.Sleep(200);
                    string text = Test_Issue242.GetWindowText(btnHandle);
                    Console.WriteLine("Control Text: " + text);
                }).Start();
            };

            var timer = new System.Windows.Forms.Timer()
            {
                Interval = 10000,
                Enabled = true
            };
            timer.Tick += (s, e) =>
            {
                timer.Enabled = false;
                button.Text = finalText;
                Debug.WriteLine("Text changed to: " + button.Text);
            };
            Application.Run(form);
        }
    }
}
```
* In `Program.cs`, set a breakpoint on the line with `string text = Test_Issue242.GetWindowText(btnHandle);`:
![grafik](https://user-images.githubusercontent.com/13289184/69876899-07f97b00-12c1-11ea-8fce-4de56b6cd3f4.png)

* Run the application with <kbd>F5</kbd>.
* Once the breakpoint is hit, within 10 seconds, press <kbd>F11</kbd> multiple times until you are past the line `int textLengthWithNullTerminator = GetWindowTextLengthW(hWnd) + 1;`.
* Note that the `textLengthWithNullTerminator` has value `2`:
![grafik](https://user-images.githubusercontent.com/13289184/69876426-d502b780-12bf-11ea-9c4e-8e16bc5d3daf.png)
* Wait a few seconds, so that at least 10 seconds have passed since the first breakpoint was hit. After that, continue by pressing <kbd>F11</kbd> until you have passed the line with `int actualTextLength = GetWindowTextW(...)`. (Before hitting that line, the debug output `Text changed to: ABC` should have been printed.)
* Note that `actualTextLength` has value `2`, and `"AB\0"` was written to the buffer:
![grafik](https://user-images.githubusercontent.com/13289184/69876556-2dd25000-12c0-11ea-8088-0661bd5f5524.png)
* Continue debugging and note that the condition `actualTextLength > textLengthWithNullTerminator` evaluates to `false` (so the branch is not entered), and eventually the function returns the value `"AB"`, rather than `"A"` or `"ABC"`, which means the condition did not work:
![grafik](https://user-images.githubusercontent.com/13289184/69876615-55291d00-12c0-11ea-8f9f-fa503f6aea92.png)

In a similar way, you can verify the first issue, e.g. by using `initialText = "AAAABBBBCCCCDDD"` and `finalText = "AAAABBBBCCCCDDDx"`. When debugging, you can see that the buffer does not contain a `\0` char, which indicates it was written past the end of the buffer:
![grafik](https://user-images.githubusercontent.com/13289184/69876978-3a0add00-12c1-11ea-964a-27b40a116573.png)

However, if you increase the number passed to `ArrayBuffer<char>.Shared.Rent`, you can see that now the `\0` char is visible at index 16:
![grafik](https://user-images.githubusercontent.com/13289184/69877040-6a527b80-12c1-11ea-8472-3425f8a782cf.png)

This shows that the buffer passed to `GetWindowTextW` was too small (16 chars instead of 17).

I have not tried to reproduce the third issue (infinite loop). It could be possible by changing the timer event code to not disable the timer, but instead alternate the text between a short and a long string value. However, it is probably much harder to reproduce.

Thanks!

**Comment by RussKie** [comment]
Awesome analysis, thank you!
Since you did all the hard work, would you like to send a PR with a fix? I promise it won't take as long as #1133 :)

**Comment by kpreisser** [comment]
> Since you did all the hard work, would you like to send a PR with a fix? I promise it won't take as long as #1133 :)

Thanks! I opened #2460.

**Comment by kpreisser** [comment]
> Additionally, all of mentioned issues will only be obversable if the text length of the window/control changes while the code is executed, which would require the window handle to belong to a different thread (or even process).

Actually, when the window/control belongs to the current process, `GetWindowTextLength` and `GetWindowTextW` cause a [`WM_GETTEXTLENGTH`](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-gettextlength) and a [`WM_GETTEXT`](https://docs.microsoft.com/en-us/windows/win32/winmsg/wm-gettext) message to be sent to the control, so a test case could change the text between those calls by overriding `WndProc` and calling `SetWindowTextW` to alter the text.

That way, the second and third issues (truncated string, infinite loop) can be reproduced. However, the first issue (out-of-bounds write) will still need to be verified manually.
